### ISSUE TITLE
How should flow analysis handle record patterns?

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#2823]
I'm working on flow analysis for record patterns right now, and it's a little weird because the required type of a record pattern isn't really the type the user intuitively expects.  Consider this code:
```dart
void handleIntPair((int, int) pair) { ... }
void f(Object x) {
  if (x case (int i, int j)) {
    print('Got integers $i and $j');
    handleIntPair(x); // (1)
  }
}
```
Intuitively it seems pretty clear that the code at (1) should be allowed, because any value `x` that matches the pattern `(int i, int j)` must have type `(int, int)`, and therefore `x` should be promoted to `(int, int)` within the scope of the `if` block.

For most of my implementation of flow analysis for patterns, I've tried to simply do the promotions that would be implied by naively desugaring the patterns into imperative code according to the rules in the spec.  But for records that doesn't work, because the required type of the pattern `(int i, int j)` isn't `(int, int)`, as you might expect.  It's `(Object?, Object?)`.

Everything works out in the end, because after we check the required type, the next step is to match the subpatterns `int i` and `int j`, so the only way for the *whole* pattern to match is if the type of `x` was actually `(int, int)`.  But it complicates flow analysis because it means that it needs extra machinery to figure out what type to promote to, even if it's not the same as the required type.

My current plan for making this all work with flow analysis is to defining a notion of a pattern's *demonstrated type*.  I won't give a detailed definition here, but the idea is that it's the most specific type that we can be certain that the scrutinee has *after* a successful match.  It's mostly the same as the required type, except that:
- It's defined for all patterns ("required type" isn't currently defined for logical-or, logical-and, relational, cast, null-check, null-assert, or constant patterns).
- It's always a subtype of the matched value type.
- It accounts for all the ways that a pattern can promote by performing a null check (including the pattern `!= null`, which doesn't promote by means of a required type).
- For record patterns, the demonstrated type is formed by combining together the demonstrated types of subpatterns.  So for example, the demonstrated type of `(int i, int j)` is `(int, int)`.

Then, the way flow analysis handles a record pattern is:
1. First, it promotes the scrutinee to the required type (`(Object?, Object?)` in this example).
2. Then, it visits all subpatterns, collecting their demonstrated types.
3. Then, it forms a record type from the demonstrated types of the subpatterns (`(int, int)` in this example), and promotes the scrutinee to that type.

Technically, since the promoted type of the scrutinee isn't observable during step 2, we could drop step 1, but I kind of like keeping it in there, because it maintains the parallel between the steps performed by flow analysis and the steps performed at runtime, and that makes it easier to reason about soundness.

So that was all background.  The reason I'm filing this issue is because, in a conversation today, @munificent suggested an alternative approach: instead of defining this new notion of "demonstrated type", what if we just make "required type" have the behaviours we want?  That would look something like this:
- The required type of a record pattern would be a record type formed from the required types of all its subpatterns.  So the required type of `(int i, int j)` would be `(int, int)` after all.
- We would have to define the required type for logical-or, logical-and, relational, cast, null-check, null-assert, and constant patterns.  This should be pretty straightforward.
- There would be a slight user-observable behaviour difference, because record types would be checked more eagerly.  For example, consider:

```dart
class C {
  bool operator <(other) {
    print('< called');
    return true;
  }
}
f() {
  C c = C();
  num n = 1.5;
  if ((c, n) case (C() && > 0, int _)) {
    print('Match succeeded');
  } else {
    print('Match failed');
  }
}
```
Today, this prints `< called`, then `Match failed`.  Under this new proposal, it would simply print `Match failed`, because it the runtime type of `(c, n)` is `(C, double)`, which is not a subtype of the required type `(C, int)`.

A drawback of this idea is that during type analysis, we wouldn't know the required type of a record pattern until *after* visiting it.  But at runtime, we have to check the required type of the record pattern as our *first* step.  This means that we would no longer be able to maintain the parallel between the steps performed by flow analysis and the steps performed at runtime.  But I don't think this would introduce any soundness problems.

Another complication is what to do for the required type of a null-assert pattern.  Intuitively it seems like it should be the matched value type promoted to non-null.  But if we do that, then the pattern `(_!,)` has the required type `(Object,)`, which means that attempting to match `(null,)` against `(_!,)` produces a match *failure*, whereas probably what the user expects is for an exception to be thrown.

CC @dart-lang/language-team

**Comment by lrhn** [comment]
I think it sounds reasonable to generalize "required type". Not sure we can use the same definition everywhere, though, because, e.g., list patterns behave differently in declaration context and match context.
(I assume the "required type" is the type used as requirement in declaration/assignment patterns, which is why it's not generalized to patterns not allowed in declarations/assignments.)

Using required type in match patterns isn't what we *generally* want, even if we might for record patterns.

For list patterns, a `var [int x] = ...` has a required type of `List<int>`, because otherwise the sub-pattern is not irrefutable against the element type (the length can still fail, but that's just list patterns being unsafe).

A `case [int x]:` can succeed with a `List<Object?>`, as long as `.single` is an integer, so the `Object?` is not so much a required type for the matched value as such, more of an early-bail-out type - if it's not a `List<Object?>`, we can reject without needing to check sub-patterns, or being able to for that matter. It's really just the first step of the pattern match, needed to succeed in order to be able to continue with further requirements. We can't use the required type here as the first guard.

Same for records at the static type level, a `var (int x, int y) = ...` *requires* a static matched value type of `(int, int)`, but `case (int x, int y) :` accepts a _static type_ of `(Object?, Object?)`. 
The difference is that it still requires a runtime type of `(int, int)` to succeed.
Records are special in that a runtime type of precisely `(Object, Object)` ensures that the field values have runtime type precisely `Object`, not `int`, where a `List<Object?>` can have elements of type `int`.

So, using the required type, as computed for declarations, as a first check in match patterns, will be incorrect.
```dart
var ([int x], [int y]) = ...; // required type (List<int>,List<int>)
...
case ([int x], [int y]): ... // Should check (List<Object?>,List<Object?>)
```

So, the algorithm for finding `List<int>` from `var [int x] = ...` can't necessarily be reused for `case [int x]:`, and that goes for nested types too. The *required type* of that is still `List<Object?>` (or `List<? super int>` if we had that notation). We can't reject early if the value is a `List<Object?>`.

I think we will need two separate definitions, "required type" for declarations, and "demonstrated type" for match patterns, because they differ at some patterns.

**Comment by eernstg** [comment]
_tl;dr  I think we can generalize 'required type', and use the field types for records  rd;lt_

I just went over the occurrences of 'required type' in the feature specification. I think we can use the following rule:

> The notion of a _required type_ is defined for a pattern that occurs in a declaration context or in an assignment context. Consider a `<patternVariableDeclaration>` of the form `var p = e;` or `final p = e;`, or a `<patternAssignment>` of the form `p = e`. In both cases, it is a compile-time error if the static type of `e` isn't assignable to the required type of `p`.

This error is already specified. The only new thing is that we're restricting the required type to be in an irrefutable context.

The above rule would be the only situation where the required type of a pattern is used. It follows that there is no need to define the required type of a refutable pattern. In particular, there is no need to do it for logical-or, relational, null-check, or constant patterns. For logical-and, it's the standard lower bound of the required types of the operands; for cast it's `Object?`; and for null-assert it's `T?`, where `T` is the required type of the subpattern (no problems with that).

The required type would otherwise be defined as it is today, except that the required type of a record pattern is obtained from the required types of its fields.

There is no need to [say](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#:~:text=Also%20variable%2C%20list%2C%20map%2C%20record%2C%20and%20object%20patterns%20only%20match%20a%20value%20of%20a%20certain%20required%20type) that 'variable, list, map, record, and object patterns only match a value of a certain required type', we just rely on the 3-phase processing to provide actual type arguments to list, map, and object patterns, and declared types to variable patterns, and we rely on subpatterns for the matching of record patterns.

In other words, the dynamic semantics should be specified in terms of the actual type arguments of the given patterns (list and map) and the declared types of the variables (i.e., variable patterns and identifier patterns, whose declared type may be specified explicitly or inferred). This means that there is no need to consider the required type of any pattern for the dynamic semantics, the required type is _only_ used to check assignability at the end of the 3-phase analysis.

---

As an aside, note that the required type of a map pattern of the form `<K, V>{...}` would be `Map<K, V>`, and also note that the type arguments would be given for all such patterns: We're in the [third phase](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#:~:text=Calculate%20the%20static%20type%20of%20the%20pattern), and if those type arguments weren't written by the developer they will have been obtained from the static type of the initializing expression (with the usual defaults, e.g., when the initializing expression has type `dynamic` then `K` and `V` are also `dynamic`) .

```dart
void main() {
  var ({"one": num? v1, ...} && map) = {"one": 1, 2: 2};
  print(map.runtimeType);
}
```

The type schema of the pattern is specified to be `Map<_, num?>`, so we'd expect the initializing expression to be inferred as `<Object, num?>{"one":1, 2: 2}`, and the map pattern would then be finalized to `<Object, num?>{"one": num? v1, ...}` whose required type is `Map<Object, num?>`, and hence the declaration has no error.

(The dynamic type of the initializing expression is actually `Map<Object, int>`, but this behavior may still be under development. The analyzer gives `map` the expected type `Map<Object, num?>`.)

**Comment by munificent** [comment]
Thanks for writing this all up! Going through all that, I think I'm leaning towards your current approach of separating required and demonstrated type too. I suspect they will diverge in some places and trying to jam them together, especially this late when we might not have time to work through all of the ramifications, could end up being regretful.

Here's an example that comes to mind:

```dart
(num, Object) record = (1, "str");
var (_ as int, _ as String) = record;
```

This program should be statically fine, which means the required type of the record pattern on the second line should be `(Object?, Object?)`.

Now consider:

```dart
(num, Object) record = (1, "str");
if (record case (_ as int, _ as String)) {
  print(record.$2.length);
}
```

I think this program should be fine too. Inside the body, we know that if the cast patterns succeeded, then the record must have a demonstrated type of `(int, String)`.

I don't see how we can merge the definition of required and demonstrated type for record patterns containing casts and have both of these work. Likewise with null-assert patterns.

I also suspect we'll run into some weird behavior around implicit casts if we aren't careful.

**Comment by eernstg** [comment]
I think we'd need a slightly more general notion (assuming that 'demonstrated type' is intended to be a property of a pattern).

The demonstration would actually be associated with the fields of an object whose type is already known to be a record type with a specific shape. In other words, it's about promotion of records, not about record patterns.

For a record `r` with a given static type (example: `(Object, num)`), promotion can take place per field. That is, a test like `r.$1 is String` can be use to promote `r` (in the example: to `(String, num)`). Promotion per field during pattern matching is just a special case, based on the fact that pattern matching can also evaluate `r.$1` using various syntaxes.

**Comment by stereotype441** [comment]
After playing with the implementation last week, I found a much simpler definition we can use for the "demonstrated type" of a pattern or subpattern: it's simply the type that flow analysis has promoted the matched value to after visiting the pattern.  In effect, this means:
- For a variable pattern, it's the type of the variable, if that's a subtype of the matched value type; otherwise it's the matched value type.
- For a cast pattern, it's the type being cast to, if that's a subtype of the matched value type; otherwise it's the matched value type
- For a constant pattern, it's the matched value type
- For a logical-and pattern, it's the demonstrated type of the RHS (which might include promotions from the LHS, since the demonstrated type of the LHS is used as the matched value type for the RHS)
- For a logical-or pattern, it's the demonstrated type of the LHS and the RHS, if those types are the same; otherwise matched value type
- For a list, map, or object pattern, it's the required type, if that's a subtype of the matched value type; otherwise it's the matched value type
- For a null-assert or null-check pattern, it's the matched value type promoted to non-nullable
- For a record pattern, it's a record type formed from the demonstrated types of the subpatterns, if that's a subtype of the matched value type; otherwise it's the matched value type
- For a relational pattern, it's the matched value type, except in the case of the pattern `!= null`, in which case it's that type promoted to non-nullable
- For a wildcard pattern, it's the explicit type, if that's a subtype of the matched value type; otherwise it's the matched value type

A nice consequence of this definition, since it's based on the promotion applied to the matched value of the subpattern, is that it gives us consistency between how an individual scrutinee is promoted and how a record is promoted.  In other words, if `if (x1 case P1)` promotes `x1` to `T1` and `if (x2 case P2)` promotes `x2` to `T2`, then `var rec = (x1, x2); if (rec case (P1, P2))` promotes `rec` to `(T1, T2)`.  Which seems like a good property.

I'm going to go with this plan (and not make any changes to "required type") unless there are objections.

**Comment by munificent** [comment]
I love it! 🚢

**Comment by eernstg** [comment]
Sounds good, @stereotype441!

Just one thing came to mind:

> - For a constant pattern, it's the matched value type

For the cases `<qualifiedName>`, `<constObjectExpression>`, and `'const' '(' <expression> ')'`, the actual action taken during matching is to call operator `==` on the constant object with the matched object as the argument.

I think it's impossible (because we don't know that we have primitive equality here) to conclude anything new about the dynamic type of the matched object based on the fact that this operator returned `true`.

**Comment by stereotype441** [comment]
> Sounds good, @stereotype441!
> 
> Just one thing came to mind:
> 
> > * For a constant pattern, it's the matched value type
> 
> For the cases `<qualifiedName>`, `<constObjectExpression>`, and `'const' '(' <expression> ')'`, the actual action taken during matching is to call operator `==` on the constant object with the matched object as the argument.
> 
> I think it's impossible (because we don't know that we have primitive equality here) to conclude anything new about the dynamic type of the matched object based on the fact that this operator returned `true`.

Agreed!  This is why I said that the demonstrated type is the *matched value* type (in other words no promotion happens).  The alternative (promoting to the *constant*'s type) would be unsound if the constant's runtime type includes an override of `operator ==` that's not well-behaved.  And there's no way to reliably know the constant's runtime type until after constant evaluation, which happens after flow analysis.

**Comment by eernstg** [comment]
Ah, of course! I misread 'For a constant pattern, it's the matched value type'.

**Comment by munificent** [comment]
Is this done now, @stereotype441? Want to do the honors and close it?

**Comment by stereotype441** [comment]
> Is this done now, @stereotype441? Want to do the honors and close it?

Yes, it's done!  Thanks for the ping 😃
